Refactor hyperparameter search backends

alexmojaki opened the pull request:
Fixes https://github.com/huggingface/transformers/issues/24379

The goal here is to clearly group the essential info/functionality about each backend together to make reading/changing things easier. For example if another backend integration is added it should be less likely for something to be forgotten as apparently happened with wandb.

@sgugger sorry I didn't get a full confirmation to go ahead with this, it just seemed easier to show what I meant with code rather than continue explaining in the issue. There's many other ways this could be done and I can change the approach but I hope that the general direction at least is clear from this PR.

I also think this would help move towards improving the user facing API since as mentioned in https://github.com/huggingface/transformers/issues/24278#issuecomment-1599189018 (cc @hugocool) the kwargs have no type hints and are not very easy to use. So maybe instead of:

```python
          best_run = trainer.hyperparameter_search(
                direction="maximize",
                backend="ray",
                # this is just **kwargs, not so clear what's possible...
                storage_path="...",
                callbacks=...,
            )  
```

one could write:

```python
          best_run = trainer.hyperparameter_search(
                direction="maximize",
                backend=RayTuneBackend(
                    # now more assistance is possible
                    storage_path="...",
                    callbacks=...,
                ),
            )  
```

alexmojaki committed:
Refactor hyperparameter search backends

sgugger commented on src/transformers/hyperparameter_search.py:
```diff
@@ -0,0 +1,128 @@
+)
+
+
+class HyperParamSearchBackendBase(ABC):
```
Let's make this a mixin but please remove all abstractmethods or the ABC superclass. We don't really use those in Transformers.

sgugger commented on src/transformers/hyperparameter_search.py:
```diff
@@ -0,0 +1,128 @@
+
+class HyperParamSearchBackendBase(ABC):
+    @abstractmethod
+    def name(self) -> str:
+        raise NotImplementedError
```
Why is this method? It's just a class attribute.

alexmojaki replied:
It was for the enforcement offered by `abc`.

sgugger commented on src/transformers/hyperparameter_search.py:
```diff
@@ -0,0 +1,128 @@
+                f"Use `pip install {self.pip_install}`."
+            )
+
+    def pip_install(self):
+        return self.name()
```
Same here, this is just a class attribute.

sgugger reviewed (commented):
Thanks for your PR, this is not what we discussed on the issue which I recommended to just complete the error message to include wandb.

Using abstract classes like this is not really the way the Transformers library is designed and we don't heavily refactor everything as you did, preferring clearer code that can be more verbose but doesn;t require the reader to jump to superclasses all the time to get the behavior of a method. I also don't see how you would have benefits for IDE as you show in the PR description.

alexmojaki committed:
Simpler refactoring without abstract base class

alexmojaki commented:
> Using abstract classes like this is not really the way the Transformers library is designed

That's fine, I was very unsure which approach to take. `abc` offers additional safety and IDE assistance as the standard way to ensure that all abstract methods are implemented, but it's probably overkill here and I also didn't like how heavy it was. I've pushed a much simpler strategy.

> I recommended to just complete the error message to include wandb.

The point of this is that it's difficult to see all the missing bits. The current code isn't just missing wandb in the error message, it's also missing from `default_hp_space` (fixed in this PR) and the docstring (not enforced in this PR, although it could be, I just didn't want to jump there just yet).

alexmojaki committed:
black

alexmojaki commented:
> I also don't see how you would have benefits for IDE as you show in the PR description.

Sorry for the confusion, that's not part of this PR to keep the scope focused, but if this is merged I can follow it up with another which adds constructors to each backend class which accept the precise kwargs that the backend `run` supports.

sgugger commented on src/transformers/hyperparameter_search.py:
```diff
@@ -0,0 +1,86 @@
+
+
+class HyperParamSearchBackendBase:
+    name: HPSearchBackend
```
Let's leave the name as string here, but have the dictionary you build later be enum->backend class.

sgugger commented on src/transformers/hyperparameter_search.py:
```diff
@@ -0,0 +1,86 @@
+class HyperParamSearchBackendBase:
+    name: HPSearchBackend
+    pip_package: str = None
+    is_available: Callable[[], bool]
+    run: Callable
+    default_hp_space: Callable
```
Let's leave these outs and implement them as methods in the subclasses.

sgugger commented on src/transformers/hyperparameter_search.py:
```diff
@@ -0,0 +1,86 @@
+
+
+class OptunaBackend(HyperParamSearchBackendBase):
+    is_available = staticmethod(is_optuna_available)
+    run = staticmethod(run_hp_search_optuna)
+    default_hp_space = staticmethod(default_hp_space_optuna)
```
Let's just implement those as methods, it's a bit more inline with the style of the lib (Here and all other classes below). Let's also add `name = "optuna"`.

sgugger commented on src/transformers/hyperparameter_search.py:
```diff
@@ -0,0 +1,86 @@
+    default_hp_space = staticmethod(default_hp_space_wandb)
+
+
+all_backends: Dict[HPSearchBackend, Type[HyperParamSearchBackendBase]] = {
```
This should be named ALL_HYPERPARAMETER_SEARCH_BACKENDS` (all caps since it's a constant and not a function). I don't think the type hint is very useful.

sgugger commented on src/transformers/hyperparameter_search.py:
```diff
@@ -0,0 +1,86 @@
+    HPSearchBackend.WANDB: WandbBackend,
+}
+
+assert list(all_backends) == list(HPSearchBackend)
```
This should be a test in the CI, not asserted at the initialization of the module like this.

sgugger commented on src/transformers/hyperparameter_search.py:
```diff
@@ -0,0 +1,86 @@
+
+assert list(all_backends) == list(HPSearchBackend)
+
+for _name, _backend in all_backends.items():
+    _backend.name = _name
```
It's not super clean to set this like that here, as mentioned before, let's just define the name as string on all backends objects.

sgugger commented on src/transformers/hyperparameter_search.py:
```diff
@@ -0,0 +1,86 @@
+
+def default_hp_search_backend() -> str:
+    available_backends = [backend for backend in all_backends.values() if backend.is_available()]
+    if available_backends:
```
We don't rely on Python bool conversion (as it doesn't work for tensors) so let's make this explicit:
```suggestion
    if len(available_backends) > 0:
```

sgugger commented on src/transformers/hyperparameter_search.py:
```diff
@@ -0,0 +1,86 @@
+def default_hp_search_backend() -> str:
+    available_backends = [backend for backend in all_backends.values() if backend.is_available()]
+    if available_backends:
+        # TODO warn if len(available_backends) > 1 ?
```
That's actually a good idea to at least put an info here and let the user know which one was auto-selected.

sgugger commented on src/transformers/trainer.py:
```diff
@@ -66,6 +57,7 @@
 from .debug_utils import DebugOption, DebugUnderflowOverflow
 from .deepspeed import deepspeed_init, deepspeed_load_checkpoint, is_deepspeed_zero3_enabled
 from .dependency_versions_check import dep_version_check
+from .hyperparameter_search import all_backends, default_hp_search_backend
```
Adapt all_backends here and below to the new name.

sgugger reviewed (commented):
Thanks for bearing with me! I have additional comments but I think this is a better fit for the style of the library already.

alexmojaki committed:
review comments:
specify name in class
use methods instead of callable class attributes
name constant better

alexmojaki committed:
review comments: safer bool checking, log multiple available backends

alexmojaki committed:
test ALL_HYPERPARAMETER_SEARCH_BACKENDS vs HPSearchBackend in unit test, not module. format with black.

alexmojaki commented:
Opened https://github.com/huggingface/huggingface_hub/issues/1526 in regards to the unrelated test failure.

sgugger commented on src/transformers/hyperparameter_search.py:
```diff
@@ -0,0 +1,121 @@
+from .integrations import (
```
Can just add a copyright here similar to all other files in the lib (potentially switching the year to 2023 if it's not)?

sgugger reviewed (approved):
Thanks for iterating! One last nit and we should be good to merge. The error in the CI is indeed unrelated.

alexmojaki committed:
copyright

sgugger commented:
Thanks a lot for your contribution!

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/transformers/pr_24384). All of your documentation changes will be reflected on that endpoint.
